compiler/native/compile.go

```diff
@@ -110,6 +196,20 @@ func (c *client) Compile(v interface{}) (*pipeline.Build, error) {
 		return nil, err
 	}
 
+	if c.ModificationService.Endpoint != "" {
+		// send config to external endpoint for modification
+		p, err = c.modifyConfig(p, build, repo)
```

Comment:
I can't remember but one thing to keep in mind is the expand process. I noticed you put this below it for steps and stages which is how I think I would do it as well. I'm thinking this might limit the administrator from being able to write modification rules for scoping the location of templates or auto injecting a template.

I'm not really sure the best way to get around that other than potentially checking in with the endpoint twice. Which doesn't sound great to my brain but I'm sure there are options. Also, wondering if you considered making this more of a first-class citizen to the interface than the approach you started with a private function.

Reply:
I also contemplated sending the config to the modification endpoint two times to allow changes pre and post template expansion. How about we add a flag on the server to allow the administrator to define if they want the processing to occur pre and post template expansion? Granted I have some concerns arise with potentially doing multiple external calls due to the [10 second](https://developer.github.com/v3/guides/best-practices-for-integrators/#favor-asynchronous-work-over-synchronous) and [30 second](https://developer.github.com/enterprise/2.21/v3/guides/best-practices-for-integrators/#favor-asynchronous-work-over-synchronous) timeout of GitHub webhooks.

Unsure adding the private function as a first-class citizen to the interface will provide much value, but happy to modify if you would prefer that.

Reply:
Yeah, the API calls are definitely a concern. I would be fine just picking pre or post and documentation what's possible for now.

I know an alternative approach to using HTTP for these types of rules is we could could up with a way to pass a RuleEngine at runtime to the compiler and then we wouldn't need HTTP at all. I'm not sure if you contemplated what that looks like, I think both patterns have pros/cons. I don't have personal experience with either approach.

Reply:
I briefly considered using a [Go Plugin](https://github.com/hashicorp/go-plugin) approach to have the server interact  with another Go binary over local GRPC for modification. Some brief thoughts on that approach:

- Added complexity for majority of administrators since they are unlikely to have the Go Plugin knowledge/background.
- Many administrators may choose to deploy the modification service alongside the server(s) so timeouts and network failures may be less concerning.
- Having the ability to deploy the modification service externally allows for constant changes without needing to redeploy the server(s).

Reply:
Cool, I think let's run with the HTTP approach then